be/src/exec/sorting/merge_path.h

```diff
@@ -0,0 +1,435 @@
+               _right_input->len == 0;
+    }
+
+private:
+    void _setup_input();
+
+    Node* _left;
+    Node* _right;
+
+    InputSegmentPtr _left_input;
+    InputSegmentPtr _right_input;
+    bool _input_ready = false;
+
+    size_t _merge_length;
+};
+
+class LeafNode final : public Node {
+public:
+    LeafNode(MergePathCascadeMerger* merger) : Node(merger) {}
+
+    void process_input(int32_t parallel_idx) override;
+    bool dependency_finished() override { return _provider_eos; }
+    bool is_leaf() override { return true; }
+    bool input_finished() override { return dependency_finished(); }
+    bool is_pending() override { return !_provider_eos && !_provider(true, nullptr, nullptr); }
+
+    void set_provider(MergePathChunkProvider&& provider) { _provider = std::move(provider); }
+
+private:
+    bool _provider_eos = false;
+    MergePathChunkProvider _provider;
+};
+
+// Transition graph is as follows:
+//                   ┌────────── PENDING ◄──────────┐
+//                   │                              │
+//                   │                              │
+//                   ├──────────────◄───────────────┤
+//                   │                              │
+//                   ▼                              │
+//       INIT ──► PREPARE ──► SPLIT_CHUNK ──► FETCH_CHUNK ──► FINISHED
+//                   ▲
+//                   |
+//                   | one traverse from leaf to root
+//                   |
+//                   ▼
+//                PROCESS
+//
+// The overall merge process will contain multiply traverses of the merge tree. And for one
+// traverse, the changelist of the stage might be:
+//                  PREPARE(level=0, leaf)
+//                     │
+//                     ▼
+//                  PROCESS(level=0)
+//                     │
+//                     ▼
+//                  PREPARE(level=1)
+//                     │
+//                     ▼
+//                  PROCESS(level=1)
+//                     │
+//                     ▼
+//                   ......
+//                     │
+//                     ▼
+//                  PREPARE(level=<root level>)
+//                     │
+//                     ▼
+//                  PROCESS(level=<root level>)
+//
+enum Stage {
+    // Some initiation work will be done in this stage, including building the cascade merge tree
+    INIT = 0,
+
+    // Before entering the PROCESS stage, we need to build relaionships between all the parallelism and
+    // all the Nodes with the specific level id
+    PREPARE = 1,
+
+    // All the parallelism will concurrently do its own job assigned at the PREPARE stage
+    PROCESS = 2,
+
+    // All the merge processing are done here, but the rootNode::_output_segments may contains chunk with size
+    // larger than the limited size required by pipeline engine, so wo need to further split it
+    SPLIT_CHUNK = 3,
```

Comment:
> For the merge tree, each MergeNode may contain one chunk. So the merge tree may buffer 2^level 's chunk?

Yes, the buffer size for each MergeNode may be `2` * `pipeline_chunk_size` * `dop`, we need to make sure that each parallelism can process enough data(at least more than `pipeline_chunk_size`)  to gain better performance.

> What if the chunk's size is less than the limited size required by pipeline engine?
In most cases, the row number of root is equals to `pipeline_chunk_size` * `dop`, so it can be divided into `dop` parts with no remains. But for the last fetch, the last chunk size may be smaller than `pipeline_chunk_size`, and it's ok.